### PR TITLE
WebUI actions

### LINKED ISSUE #436: Run and schedule actions via WebUI
https://github.com/abrt/faf/wiki/WebUI-Actions

### PR DESCRIPTION
**Pull request by mbrysa**
Closes #436.

How to test:
* `# dnf install redis`
* `# systemctl start redis`
* Build and install FAF including `faf-celery-tasks` and `faf-blueprint-celery-tasks`
* `# sudo -u faf faf-migrate-db`
* `# systemctl start faf-celery-worker` (the worker)
* `# systemctl start faf-celery-beat` (the scheduler)
* Set `everyone_is_admin = true` in `/etc/faf/plugins/web2.conf`
* Restart `httpd`
* Go to `http://localhost/faf2/celery_tasks/` (or wherever FAF is running)

Alternatively use `zeromq` instead of Redis. In that case `/etc/faf/plugins/celery_tasks.conf` needs to be reconfigured.

### CODE REVIEW

File: config/celery-beat-env.conf [hunk 1]
```diff
@@ -0,0 +1,6 @@
+CELERY_APP="pyfaf.celery_tasks.celery_app"
+CELERYD_NODES="beat"
+CELERYD_OPTS="-C -S pyfaf.celery_tasks.schedulers.DBScheduler"
+CELERYD_PID_FILE="/tmp/faf-celery-beat.pid"
```
Comment (sorki) [inline config/celery-beat-env.conf]: pid files should end up in /var/run

File: config/celery-worker-env.conf [hunk 2]
```diff
@@ -0,0 +1,6 @@
+CELERY_APP="pyfaf.celery_tasks.celery_app"
+CELERYD_NODES="worker"
+CELERYD_OPTS="-C"
+CELERYD_PID_FILE="/tmp/faf-celery-%n.pid"
```
Comment (sorki) [inline config/celery-worker-env.conf]: same here

File: faf.spec.in [hunk 3]
```diff
@@ -417,6 +425,13 @@ Requires: %{name}-fedmsg = %{faf_version}
 %description action-fedmsg-notify
 A plugin for %{name} implementing notification into Fedmsg
 
+%package action-cleanup-task-results
+Summary: %{name}'s cleanup-task-results plugin
+Requires: %{name} = %{faf_version}
+
+%description action-cleanup-task-results
+A plugin for %{name} implementing cleaup of old task results.
```
Comment (sorki) [inline faf.spec.in]: typo cleaup / cleanup

File: config/celery-beat-env.conf (hunk 1 above)
Comment (sorki) [inline config/celery-beat-env.conf]: https://fedoraproject.org/wiki/Packaging:Tmpfiles.d

File: init-scripts/faf-celery-beat.service [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+Description=FAF Celery beat
+After=network.target
+
+[Service]
+Type=forking
+User=faf
+Group=faf
+EnvironmentFile=/etc/faf/celery-beat-env.conf
+WorkingDirectory=/etc/faf
+ExecStart=/usr/bin/python2 -m celery beat \
```
Comment (sorki) [inline init-scripts/faf-celery-beat.service]: Is there a reason to use python2 instead of just python?

File: src/pyfaf/celery_tasks/__init__.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+from celery import Celery
+from celery.signals import task_postrun
+from pyfaf.storage import DatabaseFactory, TaskResult
+from pyfaf.actions import actions
+from pyfaf.config import config
+import bunch
+import sys
+import logging
+import datetime
+from cStringIO import StringIO
```
Comment (sorki) [inline src/pyfaf/celery_tasks/__init__.py]: standard lib imports first, then 3d party libs, then pyfaf imports

File: src/pyfaf/celery_tasks/__init__.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+
+@celery_app.task()
+def run_action(name, params={}, log_level=logging.DEBUG, output_length=1000):
+    db = db_factory.get_database()
+    action = actions[name]
+    cmdline = bunch.Bunch(params)
+
+    # Logger and fake stdout to capture the output of the action
+    my_stdout = StringIO()
```
Comment (sorki) [inline src/pyfaf/celery_tasks/__init__.py]: Please use capture_output context manager:

```
from pyfaf.utils.contextmanager import captured_output
with captured_output() as (cap_stdout, cap_stderr):
    <code>
```

cap_std* are StringIO objects, sys.stdout/err is saved/restored automatically.

**Comment by sorki** [comment]
- Needs documentation 
- Needs ansible support
- automatically generated field names and help texts should have first letter upcased
- list of available actions is too cluttered

Otherwise very nice!

**Comment by sorki** [comment]
Using */5 as minutes doesn't work. Not a big deal but it would be nice to have this shorthand.


> In vixie-cron, slashes can be combined with ranges to specify step values. For example, */7 in the 
> minutes field indicates every 7 minutes (see note). It is shorthand for the more verbose POSIX form 
> 5,10,15,20,25,30,35,40,45,50,55,00. POSIX does not define a use for slashes; its rationale 
> (commenting on a BSD extension) notes that the definition is based on System V format but does not exclude the possibility of extensions.

**Comment by mbrysa** [comment]
Rebased:
* pid files in `/run/faf-celery/` using tmpfiles.d
* typo fixed
* python2 -> python
* imports reordered
* added and using `capture_output_combined` context manager
* actions list split into three columns

Using */5 as minutes works for me. Could you please retest?

File: init-scripts/faf-celery-beat.service [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+[Service]
+Type=forking
+User=faf
+Group=faf
+EnvironmentFile=/etc/faf/celery-beat-env.conf
+WorkingDirectory=/etc/faf
+ExecStart=/usr/bin/python -m celery beat \
+    -A $CELERY_APP --pidfile=${CELERYD_PID_FILE} \
+    --logfile=${CELERYD_LOG_FILE} --loglevel=${CELERYD_LOG_LEVEL} \
+    --detach $CELERYD_OPTS
```
Comment (sorki) [inline init-scripts/faf-celery-beat.service]: Only works for me without --detach and without Type=forking (both are redundant)

File: faf.spec.in [hunk 8]
```diff
@@ -1133,6 +1208,22 @@ fi
...
+%{python_sitelib}/pyfaf/celery_tasks/__init__.py*
+%{python_sitelib}/pyfaf/celery_tasks/schedulers.py*
+
+%if %{with systemd}
+%files celery-tasks-systemd-services
+%{_unitdir}/faf-celery-beat.service
+%{_unitdir}/faf-celery-worker.service
+%config(noreplace) %{_sysconfdir}/faf/celery-beat-env.conf
+%config(noreplace) %{_sysconfdir}/faf/celery-worker-env.conf
+%{_tmpfilesdir}/faf-celery-tmpfiles.conf
```
Comment (sorki) [inline faf.spec.in]: Also

`%dir /run/faf-celery/`

**Comment by mbrysa** [comment]
Forking removed, `/run/faf-celery` dir added.

**Pull request merged**